File: app/src/main/java/com/dumuzeyn/mp3player/GlobalSearchEngine.java
package com.dumuzeyn.mp3player;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;

final class GlobalSearchEngine {
    GlobalSearchResult search(List<Track> tracks, List<Playlist> playlists, String query,
            int categoryLimit) {
        String normalized = Track.normalizeSearchText(query);
        if (normalized.isEmpty()) {
            return GlobalSearchResult.empty();
        }
        ArrayList<Track> songs = new ArrayList<>();
        LinkedHashMap<String, String> artists = new LinkedHashMap<>();
        LinkedHashMap<String, String> albums = new LinkedHashMap<>();
        LinkedHashMap<String, String> genres = new LinkedHashMap<>();
        ArrayList<Playlist> matchingPlaylists = new ArrayList<>();
        for (Track track : tracks) {
            if (track.normalizedSearchText.contains(normalized)) {
                addLimited(songs, track, categoryLimit);
            }
            addGroup(artists, track.artist, normalized, categoryLimit);
            addGroup(albums, track.album, normalized, categoryLimit);
            addGroup(genres, track.genre, normalized, categoryLimit);
        }
        for (Playlist playlist : playlists) {
            if (Track.normalizeSearchText(playlist.name).contains(normalized)) {
                addLimited(matchingPlaylists, playlist, categoryLimit);
            }
        }
        return new GlobalSearchResult(songs, new ArrayList<>(artists.values()),
                new ArrayList<>(albums.values()), new ArrayList<>(genres.values()),
                matchingPlaylists);
    }

    private static void addGroup(LinkedHashMap<String, String> target, String value,
            String query, int limit) {
        String key = Track.normalizeSearchText(value);
        if (!key.isEmpty() && key.contains(query)
                && (limit <= 0 || target.size() < limit)) {
            target.put(key, value);
        }
    }

    private static <T> void addLimited(List<T> target, T value, int limit) {
        if (limit <= 0 || target.size() < limit) {
            target.add(value);
        }
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/GlobalSearchEngine.kt
package com.dumuzeyn.mp3player

class GlobalSearchEngine {
    fun search(
        tracks: List<Track>,
        playlists: List<Playlist>,
        query: String?,
        categoryLimit: Int,
    ): GlobalSearchResult {
        val normalized = Track.normalizeSearchText(query)
        if (normalized.isEmpty()) return GlobalSearchResult.empty()
        val songs = ArrayList<Track>()
        val artists = LinkedHashMap<String, String>()
        val albums = LinkedHashMap<String, String>()
        val genres = LinkedHashMap<String, String>()
        val matchingPlaylists = ArrayList<Playlist>()
        tracks.forEach { track ->
            if (normalized in track.normalizedSearchText) addLimited(songs, track, categoryLimit)
            addGroup(artists, track.artist, normalized, categoryLimit)
            addGroup(albums, track.album, normalized, categoryLimit)
            addGroup(genres, track.genre, normalized, categoryLimit)
        }
        playlists.forEach { playlist ->
            if (normalized in Track.normalizeSearchText(playlist.name)) {
                addLimited(matchingPlaylists, playlist, categoryLimit)
            }
        }
        return GlobalSearchResult(
            songs,
            ArrayList(artists.values),
            ArrayList(albums.values),
            ArrayList(genres.values),
            matchingPlaylists,
        )
    }

    private fun addGroup(
        target: MutableMap<String, String>,
        value: String?,
        query: String,
        limit: Int,
    ) {
        val key = Track.normalizeSearchText(value)
        if (key.isNotEmpty() && query in key && (limit <= 0 || target.size < limit)) {
            target[key] = value.orEmpty()
        }
    }

    private fun <T> addLimited(target: MutableList<T>, value: T, limit: Int) {
        if (limit <= 0 || target.size < limit) target += value
    }
}
